serve /netboot/mac; move away from DefaultServeMux; optionally terminate TLS

colemickens opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Optional HTTPS/TLS via cert/key flags and runtime TLS support.
  * Serve static netboot content under MAC-aware paths with safe index resolution and per-machine boot triggering.
  * SSE endpoint now registered through the server mux lifecycle.
  * New machine state transition to support HTTP-based boot flows.

* **Bug Fixes / Reliability**
  * Improved startup logging and server lifecycle management for TLS and netboot serving.

* **Documentation**
  * README updated with TLS and netboot CLI options.

* **Chores**
  * Tooling/dependency updates, vendor hash and Nix module options refreshed.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

colemickens committed:
http: serve /mac/ as well

colemickens committed:
tree: don't use net/http's DefaultServeMux

colemickens committed:
main: accept tls flags, serve with TLS if they are provided

grahamc commented on http.go:
```diff
@@ -9,9 +9,18 @@ import (
+
+	server.HandleFunc("/mac/{}", func(w http.ResponseWriter, r *http.Request) {
+		fs.ServeHTTP(w, r)
+	})
```
Hmmm... I think this is good feedback. 

Making the netboot directory a flag is pretty trivial.

I don't think we do want to serve directory listings.


--

This would be a great spot to trigger some sort of state machine transition, from like http_boot to http_fetch_uki or something, as a new event for the event stream, too.

colemickens committed:
go.mod: add certwatcher; go mod tidy

colemickens committed:
main: use certwatcher to have hot reload tlsconfig

colemickens committed:
flake.lock: Update

Flake lock file updates:

• Updated input 'nixpkgs':
    'https://api.flakehub.com/f/pinned/NixOS/nixpkgs/0.1.868392%2Brev-e9f00bd893984bc8ce46c895c3bf7cac95331127/01999520-8e0b-7774-aef8-74b2a54b285c/source.tar.gz?narHash=sha256-0m27AKv6ka%2Bq270dw48KflE0LwQYrO7Fm4/2//KCVWg%3D' (2025-09-28)
  → 'https://api.flakehub.com/f/pinned/NixOS/nixpkgs/0.1.889928%2Brev-ae814fd3904b621d8ab97418f1d0f2eb0d3716f4/019a5bef-29b7-7023-9317-10e3a34588a3/source.tar.gz?narHash=sha256-YRqMDEtSMbitIMj%2BJLpheSz0pwEr0Rmy5mC7myl17xs%3D' (2025-11-05)

colemickens committed:
main: apply some coderabbit suggestion

colemickens committed:
-netboot-dir is optional and configures where to serve netboot from

colemickens committed:
http: prevent directory listing

colemickens committed:
uki state transition stuff

colemickens committed:
flake: nixos module: add options, plumb to service exec

colemickens committed:
http: specify /mac/{mac_addr}/boot.efi to make forward progress

colemickens committed:
listen on :80 and :443

grahamc commented on flake.nix:
```diff
@@ -131,6 +131,27 @@
               };
+              netbootDirectory = lib.mkOption {
+                type = lib.types.nullOr lib.types.path;
+                default = "/netboot/mac";
```
I'd rather this be null by default, this is a sorta "weird" place, and 2x weird to make it default

colemickens committed:
flake: nixos module: remove netbootDirectory default

colemickens committed:
http: s/fmt/log/g for some printfs

grahamc commented on machine.go:
```diff
@@ -119,6 +119,8 @@ func NewMachine(mac net.HardwareAddr, broker *Broker) *Machine {
 
+			{Name: "http_fetch_uki", Src: []string{"http_boot", "point_ipxe_to_http_boot"}, Dst: "http_fetch_uki"},
+
 			{Name: "os_init", Src: []string{"http_boot", "point_ipxe_to_http_boot"}, Dst: "os_init"},
```
```suggestion
			{Name: "os_init", Src: []string{"http_fetch_uki"}, Dst: "os_init"},
```

I think we now expect all transition to os_init go via http_fetch_uki.

grahamc replied:
(...and not tidying up the `Src` part of the transition graph causes reset() to be triggered more often, which is undesirable and implies the system did not follow an expected boot flow.)

colemickens committed:
http: reject netboot requests with invalid MAC address

grahamc commented on main.go:
```diff
@@ -38,6 +39,9 @@ var (
 	httpBootURLTemplate = flag.String("http-boot-url-template", "", "URL template for HTTP boot requests, like http://netboot.target/?mac={{.MAC}}")
+	tlsCertFile         = flag.String("tls-cert-file", "", "Path to TLS Certificate File")
+	tlsKeyFile          = flag.String("tls-key-file", "", "Path to TLS Key File")
+	netbootDir          = flag.String("netboot-dir", "/netboot/mac", "Path to MACs to serve for netboot")
```
```suggestion
	netbootDir          = flag.String("netboot-dir", "", "Path to MACs to serve for netboot")
```

colemickens committed:
machine: update state transition into os_init

colemickens committed:
last fixups from grahamc

colemickens committed:
fixup from code rabbit: err check for mux init

colemickens committed:
fixup state machine tests

colemickens committed:
fixup err handling condition check for parseIP
